jlog.c

```diff
@@ -2179,6 +2179,139 @@ int jlog_ctx_read_message(jlog_ctx *ctx, const jlog_id *id, jlog_message *m) {
   }
   return -1;
 }
+int jlog_ctx_bulk_read_messages(jlog_ctx *ctx, const jlog_id *id, const int count, jlog_message *m) {
+  off_t index_len;
+  u_int64_t data_off;
+  int with_lock = 0;
+  size_t hdr_size = 0;
+  uint32_t *message_disk_len;
+  int i;
+
+  if (count <= 0) {
+    return 0;
+  }
+
+ once_more_with_lock:
+
+  data_off = 0;
+
+  ctx->last_error = JLOG_ERR_SUCCESS;
+  if (ctx->context_mode != JLOG_READ)
+    SYS_FAIL(JLOG_ERR_ILLEGAL_WRITE);
+  if (id->marker < 1) {
+    SYS_FAIL(JLOG_ERR_ILLEGAL_LOGID);
+  }
+
+  __jlog_open_reader(ctx, id->log);
+  if(!ctx->data)
+    SYS_FAIL(JLOG_ERR_FILE_OPEN);
+  __jlog_open_indexer(ctx, id->log);
+  if(!ctx->index)
+    SYS_FAIL(JLOG_ERR_IDX_OPEN);
+
+  if(with_lock) {
+    if (!jlog_file_lock(ctx->index)) {
+      with_lock = 0;
+      SYS_FAIL(JLOG_ERR_LOCK);
+    }
+  }
+
+  if ((index_len = jlog_file_size(ctx->index)) == -1)
+    SYS_FAIL(JLOG_ERR_IDX_SEEK);
+  if (index_len % sizeof(u_int64_t))
+    SYS_FAIL(JLOG_ERR_IDX_CORRUPT);
+  if (id->marker * sizeof(u_int64_t) > index_len) {
+    SYS_FAIL(JLOG_ERR_ILLEGAL_LOGID);
+  }
+
+  if (!jlog_file_pread(ctx->index, &data_off, sizeof(u_int64_t),
+                       (id->marker - 1) * sizeof(u_int64_t)))
```

Comment:
Looks like we're only reading the index once, here, and assuming that subsequent messages directly follow the one we want to read... This might be OK, but I'm nervous about it, since the single-reads would read the index for each message.